src/core/ddsi/src/ddsi_raweth.c

```diff
@@ -115,22 +158,48 @@ static ssize_t ddsi_raweth_conn_write (struct ddsi_tran_conn * conn, const ddsi_
   int sendflags = 0;
   struct msghdr msg;
   struct sockaddr_ll dstaddr;
+  struct ddsi_vlan_header vhdr;
+  ddsrt_iovec_t *iov_base;
+  uint16_t vtag;
+  size_t hdrlen;
+
   assert(niov <= INT_MAX);
   memset (&dstaddr, 0, sizeof (dstaddr));
   dstaddr.sll_family = AF_PACKET;
-  dstaddr.sll_protocol = htons ((uint16_t) dst->port);
+  dstaddr.sll_protocol = htons ((uint16_t) uc->m_base.m_base.m_port);
   dstaddr.sll_ifindex = uc->m_ifindex;
   dstaddr.sll_halen = 6;
   memcpy(dstaddr.sll_addr, dst->address + 10, 6);
+
+  vtag = (uint16_t)(((dst->port >> 20) & 0xfff) + (((dst->port >> 16) & 0xe) << 12));
+
+  memcpy(vhdr.e.dmac, dstaddr.sll_addr, 6);
+  memcpy(vhdr.e.smac, uc->m_base.m_base.gv->interfaces[0].loc.address + 10, 6);
+
+  if (vtag) {
+    vhdr.e.proto = htons ((uint16_t) ETH_P_8021Q);
+    vhdr.vtag = htons (vtag);
+    vhdr.proto = htons ((uint16_t) uc->m_base.m_base.m_port);
+    hdrlen = sizeof(vhdr);
+  } else {
+    vhdr.e.proto = htons ((uint16_t) uc->m_base.m_base.m_port);
+    hdrlen = 14;
+  }
+
+  iov_base = (ddsrt_iovec_t *)(iov - 1);
```

Comment:
Just a bit of nitpicking, I'd simply use `iov[-1]` instead of `iov_base`.

I believe all the shifting is correct, but I do think it would make sense to give it a name (like `#define DDSI_TRAN_RESERVED_IOV_SLOTS 1` and then document for `ddsi_con_write` that it has this expectation, and do a static assert here just in case).

Reply:
I do not exactly understand what you mean here with a static assert?

Reply:
Just like a normal assert, except one that works at compile time. C11 has _Static_assert for this, but we emulate it by abusing array type definitions in `DDSRT_STATIC_ASSERT`.

So here, basically `DDSRT_STATIC_ASSERT(DDSI_TRAN_RESERVED_IOV_SLOTS >= 1)`. It makes the assumption explicit and helps catch trouble if someone decides to change `DDSI_TRAN_RESERVED_IOV_SLOTS`. That's all 🙂